File: growth_rate_response/global_change_growthrate.cpp
// this C++ file contains self-defined model where the reponse is weekly growth
//rate of confirmed cases. The only difference between this script and state_space_global_autocor.cpp
//is the response variable


#include <TMB.hpp>


template<class Type>
Type objective_function<Type>::operator() (){
  
  using namespace density;
  
  
  DATA_VECTOR(grorate_change); 
  
  DATA_VECTOR(temp);
  DATA_VECTOR(temp_sq);
  DATA_VECTOR(temp_lag1);
  DATA_VECTOR(temp_lag2);
  DATA_VECTOR(windsp);
  DATA_VECTOR(windsp_lag1);
  DATA_VECTOR(windsp_lag2);
  DATA_VECTOR(humidity);
  DATA_VECTOR(humidity_lag1);
  DATA_VECTOR(humidity_lag2);
  DATA_VECTOR(uv);
  DATA_VECTOR(uv_lag1);
  DATA_VECTOR(uv_lag2);
  

  DATA_VECTOR(population);
  DATA_VECTOR(dens);
  DATA_VECTOR(mage);

  DATA_VECTOR(gdp);

  DATA_VECTOR(newtests);

  
  DATA_VECTOR(mobility);
  DATA_VECTOR(mobility_lag1);
  DATA_VECTOR(mobility_lag2);
  
  DATA_VECTOR(debtrelief);
  DATA_VECTOR(debtrelief_lag1);
  DATA_VECTOR(debtrelief_lag2);
  
  DATA_VECTOR(healthinvestment);
  DATA_VECTOR(healthinvestment_lag1);
  DATA_VECTOR(healthinvestment_lag2);
  
  DATA_VECTOR(contatracing);
  DATA_VECTOR(contatracing_lag1);
  DATA_VECTOR(contatracing_lag2);
  
  DATA_VECTOR(cases_days_since_first);
  DATA_VECTOR(cases_days_since_first_sq);
  
  DATA_FACTOR(spatioLevelfactor);
  
  DATA_VECTOR(index);
  
  PARAMETER(beta0); 
  
  PARAMETER(beta_temp);
  PARAMETER(beta_temp_sq);
  PARAMETER(beta_temp_lag1);
  PARAMETER(beta_temp_lag2);
  
  PARAMETER(beta_windsp);
  PARAMETER(beta_windsp_lag1);
  PARAMETER(beta_windsp_lag2);
  
  PARAMETER(beta_humidity);
  PARAMETER(beta_humidity_lag1);
  PARAMETER(beta_humidity_lag2);
  
  PARAMETER(beta_uv);
  PARAMETER(beta_uv_lag1);
  PARAMETER(beta_uv_lag2);
  // // 
  PARAMETER(beta_population);
  PARAMETER(beta_dens);
  PARAMETER(beta_mage);
  // 
  PARAMETER(beta_gdp);
  // 
  PARAMETER(beta_newtests);
  
  PARAMETER(beta_mobility);
  PARAMETER(beta_mobility_lag1);
  PARAMETER(beta_mobility_lag2);
  
  PARAMETER(beta_debtrelief);
  PARAMETER(beta_debtrelief_lag1);
  PARAMETER(beta_debtrelief_lag2);
  
  PARAMETER(beta_healthinvestment);
  PARAMETER(beta_healthinvestment_lag1);
  PARAMETER(beta_healthinvestment_lag2);
  
  PARAMETER(beta_contatracing);
  PARAMETER(beta_contatracing_lag1);
  PARAMETER(beta_contatracing_lag2);
  
  PARAMETER(beta_days);
  PARAMETER(beta_days_sq);
  
  //random country effects
  PARAMETER_VECTOR(mu_spatioLevel);
  PARAMETER(logbeta_spatioLevel);   Type beta_spatioLevel = exp(logbeta_spatioLevel);  ADREPORT(exp(2*beta_spatioLevel));
  
  //random slopes of days
  PARAMETER_VECTOR(mu_spatioLevel_days);
  PARAMETER(logsd_randm_slop_days);   Type sd_randm_slop_days = exp(logsd_randm_slop_days);  ADREPORT(exp(2*logsd_randm_slop_days));
  
  //random slopes of squared days
  PARAMETER_VECTOR(mu_spatioLevel_sqdays);
  PARAMETER(logsd_randm_slop_sqdays);   Type sd_randm_slop_sqdays = exp(logsd_randm_slop_sqdays); ADREPORT(exp(2*logsd_randm_slop_sqdays));
  
  //random slopes on mobility
  PARAMETER_VECTOR(mu_spatioLevel_mobility); 
  PARAMETER(logsd_randm_slop_mobility);   Type sd_randm_slop_mobility= exp(logsd_randm_slop_mobility); ADREPORT(exp(2*logsd_randm_slop_mobility));
  
  //random slopes of one week lagged mobility
  PARAMETER_VECTOR(mu_spatioLevel_mobility_lag1);
  PARAMETER(logsd_randm_slop_mobility_lag1);   Type sd_randm_slop_mobility_lag1 = exp(logsd_randm_slop_mobility_lag1); ADREPORT(exp(2*logsd_randm_slop_mobility_lag1));
  
  //random slopes of two weeks lagged mobility
  PARAMETER_VECTOR(mu_spatioLevel_mobility_lag2); 
  PARAMETER(logsd_randm_slop_mobility_lag2);   Type sd_randm_slop_mobility_lag2 = exp(logsd_randm_slop_mobility_lag2); ADREPORT(exp(2*logsd_randm_slop_mobility_lag2));
  
  //random slopes on temperature
  PARAMETER_VECTOR(mu_spatioLevel_temp);
  PARAMETER(logsd_randm_slop_temp);   Type sd_randm_slop_temp = exp(logsd_randm_slop_temp); ADREPORT(exp(2*logsd_randm_slop_temp));
  
  //random slopes on squared temperature
  PARAMETER_VECTOR(mu_spatioLevel_sqtemp);
  PARAMETER(logsd_randm_slop_sqtemp);   Type sd_randm_slop_sqtemp = exp(logsd_randm_slop_sqtemp); ADREPORT(exp(2*logsd_randm_slop_temp));
  
  // parameters in ARCH(2)
  PARAMETER(logalpha0); Type alpha0 = exp(logalpha0); ADREPORT(exp(2*logalpha0));
  PARAMETER(logalpha1); Type alpha1 = exp(logalpha1); ADREPORT(exp(2*logalpha1));
  PARAMETER(logalpha2); Type alpha2 = exp(logalpha2); ADREPORT(exp(2*logalpha2));
  

  int nObs  = grorate_change.size();
  //int nspatioLevel = NLEVELS(spatioLevel);
  
  Type nll=Type(0);
  
  //contribution of random spatioLevel effects to likelihood
  if (CppAD::Variable(mu_spatioLevel(0)))
    nll-= sum(dnorm(mu_spatioLevel,Type(0),Type(1),1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_days(0)))
    nll-= sum(dnorm(mu_spatioLevel_days,Type(0),sd_randm_slop_days,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_sqdays(0)))
    nll-= sum(dnorm(mu_spatioLevel_sqdays,Type(0),sd_randm_slop_sqdays,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_mobility(0)))
    nll-= sum(dnorm(mu_spatioLevel_mobility,Type(0),sd_randm_slop_mobility,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_mobility_lag1(0)))
    nll-= sum(dnorm(mu_spatioLevel_mobility_lag1,Type(0),sd_randm_slop_mobility_lag1,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_mobility_lag2(0)))
    nll-= sum(dnorm(mu_spatioLevel_mobility_lag2,Type(0),sd_randm_slop_mobility_lag2,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_temp(0)))
    nll-= sum(dnorm(mu_spatioLevel_temp,Type(0),sd_randm_slop_temp,1)); //iid N(0,1) random year effect
  
  if (CppAD::Variable(mu_spatioLevel_sqtemp(0)))
    nll-= sum(dnorm(mu_spatioLevel_sqtemp,Type(0),sd_randm_slop_sqtemp,1)); //iid N(0,1) random year effect
  
  
  vector<Type> state(nObs);
  vector<Type> error(nObs);
  
  for (int i=0; i<nObs; i++){
    
    Type eta = beta0;
    
    if (CppAD::Variable(beta_temp))
      eta += beta_temp*temp(i);
    if (CppAD::Variable(beta_temp_sq))
      eta += beta_temp_sq*temp_sq(i);
    if (CppAD::Variable(beta_temp_lag1))
      eta += beta_temp_lag1*temp_lag1(i);
    if (CppAD::Variable(beta_temp_lag2))
      eta += beta_temp_lag2*temp_lag2(i);
    
    if (CppAD::Variable(beta_windsp))
      eta += beta_windsp*windsp(i);
    if (CppAD::Variable(beta_windsp_lag1))
      eta += beta_windsp_lag1*windsp_lag1(i);
    if (CppAD::Variable(beta_windsp_lag2))
      eta += beta_windsp_lag2*windsp_lag2(i);
    
    if (CppAD::Variable(beta_humidity))
      eta += beta_humidity*humidity(i);
    if (CppAD::Variable(beta_humidity_lag1))
      eta += beta_humidity_lag1*humidity_lag1(i);
    if (CppAD::Variable(beta_humidity_lag2))
      eta += beta_humidity_lag2*humidity_lag2(i);
    
    if (CppAD::Variable(beta_uv))
      eta += beta_uv*uv(i);
    if (CppAD::Variable(beta_uv_lag1))
      eta += beta_uv_lag1*uv_lag1(i);
    if (CppAD::Variable(beta_uv_lag2))
      eta += beta_uv_lag2*uv_lag2(i);
    
    if (CppAD::Variable(beta_population))
      eta += beta_population*population(i);
    
    if (CppAD::Variable(beta_dens))
      eta += beta_dens*dens(i);
     
    if (CppAD::Variable(beta_mage))
      eta += beta_mage*mage(i);
     
    if (CppAD::Variable(beta_gdp))
      eta += beta_gdp*gdp(i);
    
    if (CppAD::Variable(beta_newtests))
      eta += beta_newtests*newtests(i);
    
    if (CppAD::Variable(beta_mobility))
      eta += beta_mobility*mobility(i);
    if (CppAD::Variable(beta_mobility_lag1))
      eta += beta_mobility_lag1*mobility_lag1(i);
    if (CppAD::Variable(beta_mobility_lag2))
      eta += beta_mobility_lag2*mobility_lag2(i);
    
    if (CppAD::Variable(beta_mobility))
      eta += beta_mobility*mobility(i);
    if (CppAD::Variable(beta_mobility_lag1))
      eta += beta_mobility_lag1*mobility_lag1(i);
    if (CppAD::Variable(beta_mobility_lag2))
      eta += beta_mobility_lag2*mobility_lag2(i);
    
    if (CppAD::Variable(beta_debtrelief))
      eta += beta_debtrelief*debtrelief(i);
    if (CppAD::Variable(beta_debtrelief_lag1))
      eta += beta_debtrelief_lag1*debtrelief_lag1(i);
    if (CppAD::Variable(beta_debtrelief_lag2))
      eta += beta_debtrelief_lag2*debtrelief_lag2(i);
    
    if (CppAD::Variable(beta_healthinvestment))
      eta += beta_healthinvestment*healthinvestment(i);
    if (CppAD::Variable(beta_healthinvestment_lag1))
      eta += beta_healthinvestment_lag1*healthinvestment_lag1(i);
    if (CppAD::Variable(beta_healthinvestment_lag2))
      eta += beta_healthinvestment_lag2*healthinvestment_lag2(i);
    
    if (CppAD::Variable(beta_contatracing))
      eta += beta_contatracing*contatracing(i);
    if (CppAD::Variable(beta_contatracing_lag1))
      eta += beta_contatracing_lag1*contatracing_lag1(i);
    if (CppAD::Variable(beta_contatracing_lag2))
      eta += beta_contatracing_lag2*contatracing_lag2(i);
    
    if (CppAD::Variable(beta_days))
      eta += beta_days*cases_days_since_first(i); 
    
    if (CppAD::Variable(beta_days_sq))
      eta +=beta_days_sq*cases_days_since_first_sq(i); 
    
    if (CppAD::Variable(mu_spatioLevel(0)))
      eta += beta_spatioLevel*mu_spatioLevel(spatioLevelfactor(i));
    
    if (CppAD::Variable(mu_spatioLevel_days(0)))
      eta += mu_spatioLevel_days(spatioLevelfactor(i))*cases_days_since_first(i);
    
    if (CppAD::Variable(mu_spatioLevel_sqdays(0)))
      eta += mu_spatioLevel_sqdays(spatioLevelfactor(i))*cases_days_since_first_sq(i);
    
    if (CppAD::Variable(mu_spatioLevel_mobility(0)))
      eta += mu_spatioLevel_mobility(spatioLevelfactor(i))*mobility(i);
    
    if (CppAD::Variable(mu_spatioLevel_mobility_lag1(0)))
      eta += mu_spatioLevel_mobility_lag1(spatioLevelfactor(i))*mobility_lag1(i);
    
    if (CppAD::Variable(mu_spatioLevel_mobility_lag2(0)))
      eta += mu_spatioLevel_mobility_lag2(spatioLevelfactor(i))*mobility_lag2(i);
    
    if (CppAD::Variable(mu_spatioLevel_temp(0)))
      eta += mu_spatioLevel_temp(spatioLevelfactor(i))*temp(i);
    
    if (CppAD::Variable(mu_spatioLevel_sqtemp(0)))
      eta += mu_spatioLevel_sqtemp(spatioLevelfactor(i))*temp_sq(i);
    
    
    state(i) = eta;
    
    error(i) = grorate_change(i) - state(i);
    
  }
  
  
  for (int i=0; i<nObs; i++) {
    if(index(i) == 1) nll -= dnorm(error(i), Type(0), alpha0, true);
    else{
      Type sd = alpha0;
      if (CppAD::Variable(logalpha1) &  !CppAD::Variable(logalpha2))    sd = sqrt(alpha0+alpha1*pow(error(i-1),2));
      else if (CppAD::Variable(logalpha1) & CppAD::Variable(logalpha2))  sd = sqrt(alpha0+alpha1*pow(error(i-1),2)+alpha2*pow(error(i-2),2));
      nll -= dnorm(error(i), Type(0), sd, true);
    }
  }
  
  ADREPORT(state);
  ADREPORT(error);
  
  return nll;
}

File: growth_rate_response/global_change_growthrate.R

### in this script, we use the one data source that integrates severall
# data sources. see https://rs-delve.github.io/data_software/global-dataset.html

# this script contains the code dealing with raw data and
# make a data list compatible with TMB, in which change in growth rate 
# is taken as the response variable.

#load the data file
globalraw  <- read.csv("rawData/combined_dataset_latest.csv", header=T)

###############################################
#After a preliminary analysis in datapreGlobal.R,
#we decide to select a subset of variables from the raw dataset.
dateFrom <- as.integer(gsub("-", "","2020-03-01"))
dateTo <-as.integer(gsub("-", "","2020-12-31"))
library("dplyr")
library("aweek")
library("zoo")
globalsub <- globalraw %>%
  mutate(date = as.integer(gsub("-", "",DATE))) %>%
  mutate(cases_new_smoothed = zoo::rollmean(cases_new, k = 7, fill = NA))%>%
  filter(dateFrom <= date & date <= dateTo) %>%
  mutate(weekno = date2week(DATE, numeric = TRUE))  %>%
  select(country_name,DATE,weekno,
         cases_new_smoothed, 
         cases_days_since_first,
         
         stats_population, 
         stats_population_density,
         stats_median_age,
         
         stats_gdp_per_capita,
         npi_stringency_index,
         tests_new_smoothed,
         mobility_retail_recreation,
         npi_debt_relief,
         npi_contact_tracing,
         npi_healthcare_investment,
         
         weather_precipitation_mean,
         weather_humidity_mean,
         weather_sw_radiation_mean, 
         weather_temperature_mean,
         weather_wind_speed_mean)


#calculate daily growth rate change for each country
n <- length(unique(globalsub$country_name))
grorate_change <- vector()
for(i in 1:n){
  country <- unique(globalsub$country_name)[i]
  count_one <- globalsub[globalsub$country_name==country,]
  size <- nrow(count_one)
  change_one_count <- 
    (count_one$cases_new_smoothed[2:size]-count_one$cases_new_smoothed[1:(size-1)])/count_one$cases_new_smoothed[1:(size-1)]
  grorate_change <- c(grorate_change,c(NA,change_one_count))
  }

# add calculated growth rate change to the dataset
globalsub <- globalsub  %>%
             mutate(grorate_change = grorate_change )

#remove rows that contain NA and then check if the final dataframe contains NA values
globalsub <- na.omit(globalsub) 
apply(globalsub, 2, function(x) any(is.na(x)))


#calculate weekly mean of the variables
globalweek <-    with(globalsub,
                      aggregate(cbind(grorate_change,
                                      cases_new_smoothed, 
                                      cases_days_since_first,
                                      
                                      stats_population, 
                                      stats_population_density,
                                      stats_median_age,
                                      
                                      stats_gdp_per_capita,
                                      npi_stringency_index,
                                      tests_new_smoothed,
                                      mobility_retail_recreation,
                                      npi_debt_relief,
                                      npi_contact_tracing,
                                      npi_healthcare_investment,
                                      
                                      weather_precipitation_mean,
                                      weather_humidity_mean,
                                      weather_sw_radiation_mean, 
                                      weather_temperature_mean,
                                      weather_wind_speed_mean), 
                                by=list(country_name,as.factor(weekno)), mean,na.rm=TRUE))%>% 
                rename(country = Group.1, weekno = Group.2,
                stringency_index = npi_stringency_index,
                debt_relief = npi_debt_relief,
                contact_tracing = npi_contact_tracing,
                healthcare_investment = npi_healthcare_investment) %>% 
                arrange(country,weekno) 


##############################################################
##create data on the effect-lagged variables
##############################################################   
#produce dataframes for non lagged, one week lagged and two weeks lagged variables

### create indeices for the effect lagged variables

#index for no lagged effect variables
index <- rep(0,nrow(globalweek))
for(i in 2: (nrow(globalweek)-1)){
  if(!identical(globalweek$country[i],globalweek$country[i-1])){
    index[i] = 1
    index[i+1]= 1}
}
index[1]= index[2]=1

#index for one week lagged effect 
index_lag1 <- rep(0,nrow(globalweek))
for(i in 1: (nrow(globalweek)-1)){
  if(!identical(globalweek$country[i],globalweek$country[i+1])){
    index_lag1[i] = 1
    index_lag1[i+1]= 1}
}
index_lag1[1]=1
index_lag1[nrow(globalweek)]=1

#index for two weeks lagged effect 
index_lag2 <-  rep(0,nrow(globalweek))
for(i in 2: (nrow(globalweek)-1)){
  if(!identical(globalweek$country[i],globalweek$country[i+1])){
    index_lag2[i-1] = 1
    index_lag2[i] = 1
  }
}
index_lag2[nrow(globalweek)-1]=1
index_lag2[nrow(globalweek)]=1

globalweek_nolag <- globalweek %>% filter(index == 0) 
globalweek_lag1 <- globalweek[index_lag1==0,]
colnames(globalweek_lag1) <- paste(colnames(globalweek_nolag), "lag1", sep = "_")
globalweek_lag2 <- globalweek[index_lag2==0,]
colnames(globalweek_lag2) <- paste(colnames(globalweek_nolag), "lag2", sep = "_")

weekDfnew <-  cbind(globalweek_nolag,globalweek_lag1,globalweek_lag2) %>%
  arrange(country,weekno) %>%
  select(-c("country_lag1","country_lag2","weekno_lag1","weekno_lag2"))

## We exclude countries with too few data points (<5)
## and countries with poor data qualities by visual inspection
# for Japan we exclude data before march due to the outbreak 
# in cruise Diamand
countryIndx <- as.data.frame(table(weekDfnew$country)>=5)
weekDfnew <- weekDfnew %>%
  filter(country %in% rownames(countryIndx)[countryIndx$`table(weekDfnew$country) >= 5`]) %>%
  filter(!(country %in% c("Fiji","Oman"))) %>%
  filter(!(country =="Japan" & as.numeric(weekno) <= 12))

levels(weekDfnew$country)[21] = "Bolivia"
levels(weekDfnew$country)[46] = "Ivory Coast"
levels(weekDfnew$country)[91] = "Korea"
levels(weekDfnew$country)[140] = "Russian"
levels(weekDfnew$country)[182] = "Vietnam"

#creat a new vector index to indicte if the error is normal distributed or the variance of which should 
#be caculated with autoregressive conditional heteroskedasticity
indexFit <- rep(0,nrow(weekDfnew))
indexFit[1] = 1
for(i in 2: (nrow(weekDfnew)-1)){
  if(identical(weekDfnew$country[i],weekDfnew$country[i-1])){
    indexFit[i] = 0
  }else{
    indexFit[i]= 1
  }
}
indexFit  <- c(0,indexFit[1:(nrow(weekDfnew)-1)]) + indexFit 

globaldata <- cbind(weekDfnew,indexFit)


###### normalize the predictive variables
#mean normalization
mean_norm <- function(x){
  if(max(x) != min(x)) return((x- mean(x)) /(max(x)-min(x)))
  if(max(x) == min(x)) return(rep(0,length(x)))
}

globnorm <-  cbind(globaldata[,1:3],normalizefun(as.matrix(globaldata[,4:(ncol(globaldata)-1)])),globaldata[,ncol(globaldata)]) %>%
             rename("index" = "globaldata[, ncol(globaldata)]")

globnorm[globnorm == "-Inf"] <- 0
globnorm[globnorm == "Inf"] <- 0
globnorm[globnorm == "NA"] <- 0


### creat TMB formated data list
attach(globnorm)
tmb_data_growrate =
  list(
    grorate_change =globnorm$grorate_change,
    
    temp =  weather_temperature_mean,
    temp_sq =  weather_temperature_mean^2,
    temp_lag1 =  weather_temperature_mean_lag1,
    temp_lag2 = weather_temperature_mean_lag2,
    
    windsp = weather_wind_speed_mean,
    windsp_lag1 = weather_wind_speed_mean_lag1,
    windsp_lag2 = weather_wind_speed_mean_lag2,
    
    humidity = weather_humidity_mean,
    humidity_lag1 = weather_humidity_mean_lag1,
    humidity_lag2 = weather_humidity_mean_lag2,
    
    uv = weather_sw_radiation_mean,
    uv_lag1 = weather_sw_radiation_mean_lag1,
    uv_lag2 =  weather_sw_radiation_mean_lag2,
    
    population = stats_population,
    dens = stats_population_density,
    mage = stats_median_age,
    
    gdp = stats_gdp_per_capita,
    
    newtests = tests_new_smoothed,
    # 
    strindex = stringency_index,
    strindex_lag1 = stringency_index_lag1,
    strindex_lag2 = stringency_index_lag2,
    # 
    mobility = mobility_retail_recreation,
    mobility_lag1 = mobility_retail_recreation_lag1,
    mobility_lag2 = mobility_retail_recreation_lag2,
    
    debtrelief = debt_relief,
    debtrelief_lag1 = debt_relief_lag1,
    debtrelief_lag2 = debt_relief_lag2,
    
    healthinvestment = healthcare_investment,
    healthinvestment_lag1 = healthcare_investment_lag1,
    healthinvestment_lag2 = healthcare_investment_lag2,
    
    contatracing = contact_tracing,
    contatracing_lag1 = contact_tracing_lag1,
    contatracing_lag2 = contact_tracing_lag2,
    
    cases_days_since_first =  cases_days_since_first,
    cases_days_since_first_sq = cases_days_since_first^2,
    
    spatioLevelfactor = factor(globnorm$country),
    index = globnorm$index
  )
detach(globnorm)




